### DISCUSSION TITLE
Schema definition with zod

### DISCUSSION THREAD

**Discussion by stefaanMLB** [mikro-orm/mikro-orm#5102] [Ideas]
When using zod for data validation, one still has to create every type twice: once in zod, once for mikro-orm
It would be great if the entities could be defined by extending zod schemas.

Unfortunately, Typescript doesn't allow to add decorators to elsewhere-defined classes at present.

**Comment by B4nan** [comment]
I don't see how this could work, you won't be able to represent everything the ORM offers via zod schema. Only thing that makes sense to me is the opposite direction, having a type helper which creates a zod schema from the ORM entity definition. And I already tried to hack that myself without luck, it would be pretty hard task to deliver this on type level. To be absolutely honest, I am not sure how much value in this approach even is, the ORM already validates the data, you will get errors for missing required properties, as well as for type mismatch, so the zod schema is not adding any value unless you start using more complex validations (e.g. this string is email, or this number is between 18 and 99).

> Unfortunately, Typescript doesn't allow to add decorators to elsewhere-defined classes at present.

Not sure what you mean by that? You don't need decorators to define entities, they are optional.

**Reply by darkbasic** [reply]
> I don't see how this could work, you won't be able to represent everything the ORM offers via zod schema. Only thing that makes sense to me is the opposite direction, having a type helper which creates a zod schema from the ORM entity definition.

That's exactly how it should be done!

> And I already tried to hack that myself without luck, it would be pretty hard task to deliver this on type level.

I've already implemented it and I'm using it in my application! I've briefly mentioned it in my Accounts.js blog post and I'd like to release everything as open source but I'm currently stuck on this blocker: https://github.com/colinhacks/zod/pull/2234
If you can somehow put some pressure to get this reviewed/merged I would happily release my library.

More info about my use case in the intro of the article: https://twitter.com/NiccoloBelli/status/1744315549813801030

**Comment by stefaanMLB** [comment]
The way I see it (web backend development - I use NestJS), you really want to start with defining the DTO's which are tightly connected to the business logic.  I use zod for that because it hands me DTO validation for free (next to classes and types).

Further down the road, you'll need to define entities that resemble the DTO's to some extent (more or less, depending on the situation, I am stuck with historic database schemas).  At present, that's not a very DRY workflow, the entities ressemble the DTO's , so being able to extend the entities from the DTO'd would be nice.  Doing it the other way around seems to neagte the natural workflow.
Obviously you'll need some extended zod functionality to pull this one off, but I think the solution would be unique.

I'll have a look at the possibilities to define the entities without using decorators.  I'm clearly missing out on something there.

**Reply by B4nan** [reply]
Well, but the entity needs to be the source of truth, not the DTO. You need something that represents the db table, not the (possibly many different) DTOs. This really needs to go the other way around, or it will be very limited.

> Doing it the other way around seems to neagte the natural workflow.

And doing it your way completely negates the fact what defines all the constraints - the db table (which is represented by the entity), not a DTO that is destined to define only user/API inputs of some specific endpoint.

> I'll have a look at the possibilities to define the entities without using decorators. I'm clearly missing out on something there. 

https://mikro-orm.io/docs/entity-schema

(also all examples in https://mikro-orm.io/docs/defining-entities have EntitySchema version too)

**Comment by stefaanMLB** [comment]
Hmm, this discussion centers on which objects should be the source of thruth.
I can see where the entities are tempting when you have full control over the DB schema.  But you may not have that in real-like situations.  I never have thas luxury in my work, the schema's are >20 years old and often absurdly bad (not how I would do it for sure).

I can agree that the DTO's are not the answer either in most situations.
Maybe the source must be business related objects and both the entities and the dto's should descend from there.

@darkbasic, I'll have a look at your blog and the EntitySchema stuff.  Thanks for the hints.

**Comment by stefaanMLB** [comment]
I refactored an example to `EntitySchema` which was **educational** but still leaves me with my (somewhat altered) feature request :

> Can the `EntitySchema` definition pick up the known specifications from `zod` and leave the rest to be defined in the EntitySchema constructor ?

e.g. When one defines `zodSchema = z.object({  login: z.string().length(60) })` and `ormSchema = new EntitySchema<...>(properties: {login: { type: 'string', length: 60, primary: true }})`, can the length and type properties be copied over from the business object defined in zod into the EntitySchema definition without the need to duplicate the `string` and `length` (`nullable`, `optional`, `enum`, `defaul`t, ...) parts?


The below are unfortunate facts-of-life for me (and presumably for many other developers)
- prop/column names often get translated between the DB and the business object
- type may change as well (eg '`J`'/'`N`' to/from `boolean`)
- The misalignment between the `Doctor` and `GlimsDoctor` objects is due to some legacy software we have to live with

I add the full example for reference. 
### business object definition
```ts
import { z } from 'nestjs-zod/z'
import { createZodDto } from 'nestjs-zod'

export const GlimsDoctorSchema = z.object({
  isGlimsImport: z.boolean().optional(),
  code: z.string().length(10),
  firstName: z.string(),
  lastName: z.string(),
  fullName: z.string(),
  rizivNr: z.string().optional(),
  ordeNr: z.string().optional(),
  type: DoctorTypeSchema.default('D'),
})

export type IGlimsDoctor = z.infer<typeof GlimsDoctorSchema>
export class GlimsDoctor extends createZodDto(GlimsDoctorSchema) {}

export const DoctorSchema = GlimsDoctorSchema.omit({ fullName: true, rizivNr: true, ordeNr: true })
  .extend({
    title: z.string().length(20).optional().default('Dr.'),
    login: z.string().length(60),
    passwordAvailable: z.boolean().default(false),
    password: z.string().optional(),
    location: LocationSchema,
    language: LanguageSchema,
    email: z.string(),
    riziv: z.string(),
    active: z.boolean(),
    practiceCode: z.string().optional(),
    practiceName: z.string().optional(),
  })
  .strict()

export type IDoctor = z.infer<typeof DoctorSchema>
export class Doctor extends createZodDto(DoctorSchema) {
  get fullName() {
    return `${this.firstName} ${this.lastName}`
  }
  get fullId() {
    return `${this.fullName} (${this.code})`
  }
}

export const DoctorUpdateSchema = DoctorSchema.omit({ login: true })
  .partial()
  .extend({ login: z.string() })

export type DoctorUpdate = z.infer<typeof DoctorUpdateSchema>
export class DoctorUpdateDto extends createZodDto(DoctorUpdateSchema) {}

// enum declaraties
export const locationValues = ['Kortrijk', 'Medlab', 'Ieper'] as const
export const LocationSchema = z.enum(locationValues)
export type Location = z.infer<typeof LocationSchema>

export const languageValues = ['NL', 'FR', 'EN', 'DE'] as const
export const LanguageSchema = z.enum(languageValues)
export type Language = z.infer<typeof LanguageSchema>

export const doctorTypeValues = ['D', 'V', 'A', 'R'] as const
export const DoctorTypeSchema = z.enum(doctorTypeValues)
export type DoctorType = z.infer<typeof DoctorTypeSchema>
```
### Entity definition
```ts
import { EntitySchema, OptionalProps } from '@mikro-orm/core'
import { Doctor, doctorTypeValues, languageValues, locationValues } from '@src/dtos/doctor.dto'

export class UserEntity extends Doctor {
  [OptionalProps]?: 'password' | 'firstName' | 'lastName' | 'practiceCode' | 'practiceName' | 'location' | 'title' | 'riziv' | 'location' | 'language' | 'type' | 'actief' | 'emailadres' | 'token' | 'info' | 'forwardToElab' | 'encryptedPassword' | 'fullName' | 'fullId' | 'password' | 'email' | 'password_info' | 'active' | 'isGlimsImport' | 'passwordAvailable'

  get actief() {
    return this.active ? 'J' : 'N'
  }
  set actief(value: 'J' | 'N') {
    this.active = value === 'J'
  }
  encryptedPassword?: string
  password_info = ''
  token = ''
  forwardToElab = true
  practiceCode?: string
  practiceName?: string

  [OptionalProps]?: 'password' | 'firstName' | 'lastName' | 'practiceCode' | 'practiceName' | 'location' | 'title' | 'riziv' | 'location' | 'language' | 'type' | 'actief' | 'emailadres' | 'token' | 'info' | 'forwardToElab' | 'encryptedPassword' | 'fullName' | 'fullId' | 'password' | 'email' | 'password_info' | 'active' | 'isGlimsImport' | 'passwordAvailable'}

export const UserEntitySchema = new EntitySchema<UserEntity>({
  class: UserEntity,
  tableName: 'users',
  properties: {
    login: { type: 'string', length: 60, primary: true },
    code: { type: 'string', length: 10 },
    firstName: { type: 'string', length: 35, fieldName: 'firstname', default: '' },
    lastName: { type: 'string', length: 35, fieldName: 'name', default: '' },
    type: { enum: true, items: () => doctorTypeValues, length: 1, default: 'D' },
    title: { type: 'string', length: 20, default: 'Dr.' },
    token: { type: 'string', length: 64, nullable: true, default: '' },
    password: { type: 'string', length: 15, nullable: true, default: '' },
    password_info: { type: 'string', length: 64, fieldName: 'info', default: '' },
    encryptedPassword: { type: 'string', length: 64, nullable: true, default: null },
    location: {
      enum: true,
      items: () => locationValues,
      length: 20,
      default: 'Kortrijk',
      fieldName: 'locatie',
    },
    language: {
      enum: true,
      items: () => languageValues,
      length: 2,
      default: 'NL',
    },
    email: { type: 'string', length: 100, fieldName: 'emailadres', default: '' },
    riziv: { type: 'string', length: 15 },
    actief: { enum: true, items: () => ['J', 'N'], length: 1, default: 'J' },
    active: { type: 'method', persist: false, getter: true, setter: true },
    forwardToElab: { type: 'boolean', default: true },
    practiceCode: { type: 'string', length: 20, nullable: true, fieldName: 'praktijkCode' },
    practiceName: { type: 'string', length: 50, nullable: true, fieldName: 'praktijkNaam' },
  },
})
```

**Reply by B4nan** [reply]
I am personally not interested in implementing that, as it does not feel right, the entity should be always the source of truth. To me, it feels easier to improve the EntitySchema class for added inference and allow more fine grained validation (to match what zod offers) rather than this. And it would still work the other way around, you define entity and infer a DTO schema out of it.

That does not mean it cannot be born as a 3rd party package (and I'll be happy to add more extension points if needed so it can be implemented), it's just not on my own roadmap. Such package could later on be added to the ORM too if it would make sense.

**Reply by stefaanMLB** [reply]
OK, fair enough.  However
> the entity `cannot` be the source of truth

For projects where both the back- and frontend are written in typescript you'll want to **share** at least the DTO definitions, possible also some/most of the business objects **between the front- and backend**.
To keep things dry, you'll want to define the business objects once and deduce the DTO's and entities from there.
If the entities are the source of truth this means shipping mikro-orm to the frontend.

Having zod in the front-end is fine, having mikro-orm in the frontend is not.

**Reply by B4nan** [reply]
Sorry, but you won't change my mind on this one, the entity definition represents the database table, and that is the source of truth, not DTO. This has nothing to do with how you write your project, this is about data being the source of truth, and everything else is an abstraction in the end.

**Reply by B4nan** [reply]
> If the entities are the source of truth this means shipping mikro-orm to the frontend. 

This is also weird way of thinking. You don't work with database tables on frontend, you work with whatever your API exposes - those are DTOs, not entities, they won't share the shape most of the time (at least they shouldn't, it of course depends on who is designing things).

With that said, I understand what you want to solve and why, but saying a DTO should be the single source of truth is simply not true. Few years ago I was considering to create some frontend companion lib that would allow working with the DTOs on frontend, maybe having a zod schema code generator would be the simplest way out of this, it would be pretty simple to implement too.

**Reply by stefaanMLB** [reply]
I didn't say DTO's must be the single source of truth I'm  saying the business objects should be.
In ideal cases, the `entities`, `business objects` and `DTOs` will be equal or at very similar.  If I had the choice my Doctor object would be exactly the same in the database, the backend and the frontend.  Any Doctor-related DTO would look very much like the Doctor business object.

If you define your business objects in mikro-orm, deduce DTO's from there and share DTO's between back- and frontend you'll have to ship mikro-orm to the frontend.

**Reply by dylan1951** [reply]
I have a monorepo with a separate frontend and backend. I want to use oRPC with a contract-first approach for e2e type safety and I feel like I should put the DTO's in a shared package with the API contract. But then I would end up with zod schemas that look a lot like a duplication of all my `defineEntity` schema definitions.

This idea of deriving the DTOs and DB entities from common "business objects" is interesting
